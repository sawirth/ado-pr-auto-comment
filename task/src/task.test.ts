import tl = require('azure-pipelines-task-lib/task');
import { WebApi } from 'azure-devops-node-api';
import executeTask from './task';

jest.mock('azure-devops-node-api');
jest.mock('azure-pipelines-task-lib/task');

const gitApiMock = {
    getThreads: jest.fn(),
    createThread: jest.fn(),
};
(WebApi as any).mockImplementation(() => ({ getGitApi: () => gitApiMock }));

jest.setTimeout(10000000);

describe('Integration Test', () => {
    beforeEach(() => {
        jest.clearAllMocks();

        // Mock Inputs
        (tl.getInput as any).mockReturnValueOnce('some-token');
        process.env.SYSTEM_TEAMPROJECT = 'team-project';
        process.env.BUILD_REPOSITORY_NAME = 'some-repo';
        process.env.SYSTEM_PULLREQUEST_PULLREQUESTID = '123456';
        process.env.SYSTEM_TEAMFOUNDATIONCOLLECTIONURI = 'some-org-url';
    })

    it('should fail because no comments were provided', async () => {
        (tl.getInput as any)
            .mockReturnValueOnce(null)
        gitApiMock.getThreads.mockReturnValueOnce([]);

        await executeTask();

        expect(tl.setResult).toMatchSnapshot();
        expect(gitApiMock.createThread).toMatchSnapshot();
    });

    it('should skip comments that are just whitespace', async () => {
        (tl.getInput as any)
            .mockReturnValueOnce("comment one\n   \ncomment two")
        gitApiMock.getThreads.mockReturnValueOnce([]);

        await executeTask();

        expect(tl.setResult).toMatchSnapshot();
        expect(gitApiMock.createThread).toMatchSnapshot();
    });

    it('should add comments to a pr that has none', async () => {
        (tl.getInput as any)
            .mockReturnValueOnce("comment one\ncomment two")
        gitApiMock.getThreads.mockReturnValueOnce([]);

        await executeTask();

        expect(tl.setResult).toMatchSnapshot();
        expect(gitApiMock.createThread).toMatchSnapshot();
    });

    it('should add comments to a pr that already has some', async () => {
        (tl.getInput as any)
            .mockReturnValueOnce("new comment one\nnew comment two")
        gitApiMock.getThreads.mockReturnValueOnce([
            { comments: [{content: "existing comment one"}] },
            { comments: [{content: "existing comment two"}] },
        ]);

        await executeTask();

        expect(tl.setResult).toMatchSnapshot();
        expect(gitApiMock.createThread).toMatchSnapshot();
    });

    it('should ignore a comment that already exists and add the rest', async () => {
        (tl.getInput as any)
            .mockReturnValueOnce("existing comment one\nnew comment two")
        gitApiMock.getThreads.mockReturnValueOnce([
            { comments: [{content: "existing comment one"}] },
            { comments: [{content: "existing comment two"}] },
        ]);

        await executeTask();

        expect(tl.setResult).toMatchSnapshot();
        expect(gitApiMock.createThread).toMatchSnapshot();
    });

    it('should use provided newCommentDelimiter to split comments', async () => {
        const newCommentDelimiter = "||";

        (tl.getInput as any)
            .mockReturnValueOnce(`comment one${newCommentDelimiter}comment   \n  two${newCommentDelimiter}comment\n\nthree`)
            .mockReturnValueOnce(newCommentDelimiter);

        await executeTask();

        expect(tl.setResult).toMatchSnapshot();
        expect(gitApiMock.createThread).toMatchSnapshot();
    });
});
