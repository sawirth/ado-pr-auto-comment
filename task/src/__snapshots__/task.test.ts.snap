// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Test should add comments to a pr that already has some 1`] = `
[MockFunction] {
  "calls": [
    [
      0,
      "All comments have been added",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add comments to a pr that already has some 2`] = `
[MockFunction] {
  "calls": [
    [
      {
        "comments": [
          {
            "content": "new comment one",
          },
        ],
        "status": 1,
      },
      "some-repo",
      123456,
      "team-project",
    ],
    [
      {
        "comments": [
          {
            "content": "new comment two",
          },
        ],
        "status": 1,
      },
      "some-repo",
      123456,
      "team-project",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add comments to a pr that has none 1`] = `
[MockFunction] {
  "calls": [
    [
      0,
      "All comments have been added",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add comments to a pr that has none 2`] = `
[MockFunction] {
  "calls": [
    [
      {
        "comments": [
          {
            "content": "comment one",
          },
        ],
        "status": 1,
      },
      "some-repo",
      123456,
      "team-project",
    ],
    [
      {
        "comments": [
          {
            "content": "comment two",
          },
        ],
        "status": 1,
      },
      "some-repo",
      123456,
      "team-project",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should fail because no comments were provided 1`] = `
[MockFunction] {
  "calls": [
    [
      2,
      "comments must be provided",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should fail because no comments were provided 2`] = `[MockFunction]`;

exports[`Integration Test should ignore a comment that already exists and add the rest 1`] = `
[MockFunction] {
  "calls": [
    [
      0,
      "All comments have been added",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should ignore a comment that already exists and add the rest 2`] = `
[MockFunction] {
  "calls": [
    [
      {
        "comments": [
          {
            "content": "new comment two",
          },
        ],
        "status": 1,
      },
      "some-repo",
      123456,
      "team-project",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should skip comments that are just whitespace 1`] = `
[MockFunction] {
  "calls": [
    [
      0,
      "All comments have been added",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should skip comments that are just whitespace 2`] = `
[MockFunction] {
  "calls": [
    [
      {
        "comments": [
          {
            "content": "comment one",
          },
        ],
        "status": 1,
      },
      "some-repo",
      123456,
      "team-project",
    ],
    [
      {
        "comments": [
          {
            "content": "comment two",
          },
        ],
        "status": 1,
      },
      "some-repo",
      123456,
      "team-project",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should use provided newCommentDelimiter to split comments 1`] = `
[MockFunction] {
  "calls": [
    [
      0,
      "All comments have been added",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should use provided newCommentDelimiter to split comments 2`] = `
[MockFunction] {
  "calls": [
    [
      {
        "comments": [
          {
            "content": "comment one",
          },
        ],
        "status": 1,
      },
      "some-repo",
      123456,
      "team-project",
    ],
    [
      {
        "comments": [
          {
            "content": "comment   
  two",
          },
        ],
        "status": 1,
      },
      "some-repo",
      123456,
      "team-project",
    ],
    [
      {
        "comments": [
          {
            "content": "comment

three",
          },
        ],
        "status": 1,
      },
      "some-repo",
      123456,
      "team-project",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
